File: packages/core/src/db/vectorStore.ts
import * as lancedb from "@lancedb/lancedb";
import OpenAI from "openai";

const TABLE_NAME = "knowledge";

export interface Document {
  id: string;
  source: string;
  content: string;
  vector: number[];
}

export interface VectorStoreConfig {
  dbPath: string;
  openaiApiKey: string;
}

export class VectorStore {
  private config: VectorStoreConfig;
  private db: lancedb.Connection | null = null;
  private table: lancedb.Table | null = null;
  private openai: OpenAI;

  constructor(config: VectorStoreConfig) {
    this.config = config;
    this.openai = new OpenAI({ apiKey: config.openaiApiKey });
  }

  private async getDB() {
    if (!this.db) this.db = await lancedb.connect(this.config.dbPath);
    return this.db;
  }

  private async getTable() {
    if (this.table) return this.table;
    const conn = await this.getDB();
    const names = await conn.tableNames();
    if (!names.includes(TABLE_NAME))
      throw new Error("No documents loaded. Run ingest first.");
    this.table = await conn.openTable(TABLE_NAME);
    return this.table;
  }

  async embedText(text: string): Promise<number[]> {
    const res = await this.openai.embeddings.create({
      model: "text-embedding-3-small",
      input: text,
    });
    return res.data[0]!.embedding;
  }

  async addDocuments(docs: Omit<Document, "vector">[]) {
    const withVectors: Document[] = await Promise.all(
      docs.map(async (d) => ({ ...d, vector: await this.embedText(d.content) }))
    );
    const conn = await this.getDB();
    const names = await conn.tableNames();
    if (names.includes(TABLE_NAME)) {
      const t = await conn.openTable(TABLE_NAME);
      await t.add(withVectors);
      this.table = t;
    } else {
      this.table = await conn.createTable(TABLE_NAME, withVectors, { mode: "create" });
    }
    console.log(`Added ${withVectors.length} chunks to vector store.`);
  }

  async search(query: string, limit = 5): Promise<Document[]> {
    const t = await this.getTable();
    const queryVec = await this.embedText(query);
    const results = await t.vectorSearch(queryVec).limit(limit).toArray();
    return results as Document[];
  }
}

File: knohub/db/vector_store.py
from __future__ import annotations

import asyncio
from dataclasses import dataclass, field
from typing import List, Optional

import lancedb
import openai
import pyarrow as pa


TABLE_NAME = "documents"
EMBEDDING_MODEL = "text-embedding-3-small"
DIMS = 1536


@dataclass
class Document:
    id: str
    text: str
    source: str
    vector: List[float] = field(default_factory=list)


class VectorStore:
    def __init__(self, db_path: str, openai_api_key: str) -> None:
        self._db_path = db_path
        self._client = openai.AsyncOpenAI(api_key=openai_api_key)

    def _get_db(self) -> lancedb.DBConnection:
        return lancedb.connect(self._db_path)

    async def embed_text(self, text: str) -> List[float]:
        response = await self._client.embeddings.create(
            model=EMBEDDING_MODEL,
            input=text,
        )
        return response.data[0].embedding

    def _add_documents_sync(self, rows: list, schema: pa.Schema) -> None:
        db = self._get_db()
        table_names = db.list_tables()
        if TABLE_NAME in table_names:
            tbl = db.open_table(TABLE_NAME)
            tbl.add(rows)
        else:
            db.create_table(TABLE_NAME, data=rows, schema=schema)

    async def add_documents(self, docs: List[Document]) -> None:
        if not docs:
            return

        schema = pa.schema([
            pa.field("id", pa.string()),
            pa.field("text", pa.string()),
            pa.field("source", pa.string()),
            pa.field("vector", pa.list_(pa.float32(), DIMS)),
        ])

        rows = []
        for doc in docs:
            vector = await self.embed_text(doc.text)
            rows.append({
                "id": doc.id,
                "text": doc.text,
                "source": doc.source,
                "vector": vector,
            })

        await asyncio.to_thread(self._add_documents_sync, rows, schema)

    def _search_sync(self, vector: List[float], limit: int) -> List[Document]:
        db = self._get_db()
        if TABLE_NAME not in db.list_tables():
            return []
        tbl = db.open_table(TABLE_NAME)
        results = tbl.search(vector).limit(limit).to_list()
        return [
            Document(id=r["id"], text=r["text"], source=r["source"], vector=list(r["vector"]))
            for r in results
        ]

    def _list_sources_sync(self) -> List[dict]:
        db = self._get_db()
        if TABLE_NAME not in db.list_tables():
            return []
        tbl = db.open_table(TABLE_NAME)
        rows = tbl.to_pandas()[["id", "source"]].to_dict("records")
        # Count chunks per source
        counts: dict = {}
        for r in rows:
            counts[r["source"]] = counts.get(r["source"], 0) + 1
        return [{"source": s, "chunks": n} for s, n in sorted(counts.items())]

    async def list_sources(self) -> List[dict]:
        return await asyncio.to_thread(self._list_sources_sync)

    async def search(self, query: str, limit: int = 5) -> List[Document]:
        vector = await self.embed_text(query)
        return await asyncio.to_thread(self._search_sync, vector, limit)
